### ISSUE TITLE
Performance Optimizations

### ISSUE THREAD

**Issue by VincentGourbin** [VincentGourbin/flux-2-swift-mlx#1]
## Overview

Current generation performance is slow and needs optimization.

### Current Performance
| Resolution | Steps | Time |
|------------|-------|------|
| 256×256 | 10 | ~20 min |
| 512×512 | 20 | ~1.7 hours |

### Optimization Targets

- [ ] **MLXFast.scaledDotProductAttention** - Use MLX's optimized attention kernel
- [ ] **Memory management** - Clear GPU cache periodically during generation
- [ ] **Profiling** - Use Instruments to identify bottlenecks
- [ ] **Batched operations** - Reduce kernel launch overhead
- [ ] **Metal shader optimization** - Investigate custom Metal kernels if needed

### Expected Improvement

Target: 5-10x speedup for a reasonable user experience.

### References

- [MLX Performance Guide](https://ml-explore.github.io/mlx/build/html/usage/performance.html)
- [mflux implementation](https://github.com/filipstrand/mflux) for performance comparison

**Comment by VincentGourbin** [comment]
## Performance & Quality Progress Report

Branch: `feature/quality-and-performance-improvements`

### 🔧 Critical Quality Fixes

Found and fixed two bugs causing poor image quality (noisy/grainy output):

1. **VAE Scaling Factor Bug**: Flux.2 VAE does NOT use the SD scaling factor (0.18215). Removed the incorrect division in `decode()`.
2. **BatchNorm Epsilon Bug**: Fixed eps from `1e-5` to `1e-4` (Flux.2 uses `batch_norm_eps = 0.0001`).

**Result**: Image file size went from 17KB (mostly noise) to 1.7MB (actual detail) for the same parameters.

### ⚡ Performance Improvements

| Metric | Before | After | Improvement |
|--------|--------|-------|-------------|
| Total time (1024x1024, 28 steps) | 42m 56s | 34m 51s | **-19%** |
| Average per step | 1m 30s | 1m 13s | **-19%** |
| Slowest step | 4m 40s | 2m 49s | **-40%** |
| Post-processing | ~15s | 1.5ms | **~10,000x** |

Implemented:
- `Flux2Profiler` for detailed timing breakdown
- Bulk GPU array extraction in post-processing
- Pre-computed guidance tensor
- Checkpoint interval support for intermediate images

### 📊 Profiler Output Example

```
╔══════════════════════════════════════════════════════════════╗
║                  FLUX.2 PERFORMANCE REPORT                   ║
╠══════════════════════════════════════════════════════════════╣
📊 PHASE TIMINGS:
  1. Load Text Encoder                4.08s    0.2% 
  2. Text Encoding                    2.41s    0.1% 
  3. Unload Text Encoder            113.7ms    0.0% 
  4. Load Transformer                23.63s    1.1% 
  5. Load VAE                        80.9ms    0.0% 
  6. Denoising Loop               34m 19.2s   98.5% ███████████████████
  7. VAE Decode                       1.94s    0.1% 
  8. Post-processing                  1.5ms    0.0% 

📈 DENOISING STEP STATISTICS:
  Steps:              28
  Average per step:   1m 13.1s
  Fastest step:       52.02s
  Slowest step:       2m 49.3s
╚══════════════════════════════════════════════════════════════╝
```

### 🎨 Bonus: Prompt Upsampling

Integrated MistralCore's prompt enhancement feature:
- `--upsample-prompt` CLI flag
- Uses BFL's official system message to enhance prompts with visual details
- Testing in progress with same prompt to compare quality

### 📁 Test Artifacts

- `cat_beach_fixed.png` - Final image (1.7MB)
- `cat_beach_fixed_checkpoints/` - Intermediate images at steps 7, 14, 21, 28

### ⏳ Next Steps

- [ ] Test with `--upsample-prompt` to compare quality
- [ ] Validate results before merging
- [ ] Consider additional optimizations (MLXFast attention, batch processing)

---
*Work in progress - do not merge yet*
